### DISCUSSION TITLE
Inactive Subdomain with no variables and kernels

### DISCUSSION THREAD

**Discussion by jmeier** [idaholab/moose#28250] [Q&A Modules: Solid mechanics] [answered]
Dear Moose Community,

I have created a very simple Moose model in which the elements of a subdomain are to be moved to an “inactive” subdomain. Following the recommendation in #28092, I tried to block-restrict all variables and kernels away from my inactive subdomain. When I run this model, Moose will not converge:

```
Time Step 3, time = 0.101, dt = 0.001
 0 Nonlinear |R| = 4.150026e+01
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_NUMERIC_ZEROPIVOT 
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge
```

When I add some MatDiffusion kernels, the model converges.

My questions are:

- How can I check which variables and kernels are active on which subdomains?
- A nuance asked differently: How can I see what is active (kernels, variables, ...) on a given subdomain?
- The variables disp_x, disp_y, and disp_z seem to be present in my inactive domain. Is there my problem? If yes, should I block-restrict them as well? 

<details><summary> Moose Model without Variables and Kernels in the inactive subdomain - but failing </summary>
<p>

```
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  use_displaced_mesh = false
[]

Box1_inactive_name = 'Box1_inactive'
inactive_domain_block_names = '${Box1_inactive_name}'

[Problem]
  solve = true
  kernel_coverage_check = SKIP_LIST
  kernel_coverage_block_list = '${inactive_domain_block_names}'
  material_coverage_check = SKIP_LIST
  material_coverage_block_list = '${inactive_domain_block_names}'
[]

[Mesh]
  [BaseMesh]
    type = GeneratedMeshGenerator
    subdomain_name = 'BaseMesh'
    elem_type = 'HEX8'
    dim = 3
    nx = 6
    ny = 6
    nz = 2
    xmin = -3
    xmax = +3
    ymin = -3
    ymax = +3
    zmin = -2
    zmax = +2
  []

  [Box1]
    type = SubdomainBoundingBoxGenerator
    block_name = 'Box1'
    input = "BaseMesh"
    block_id = 1
    location = "INSIDE"
    bottom_left = "-1.0 -1.0 -0.0"
    top_right = "+1.0 +1.0 +2.0"
  []

  add_subdomain_names = '${inactive_domain_block_names}'
[]

[Physics]
  [SolidMechanics]
    [QuasiStatic]
      [all]
        strain = SMALL
        incremental = true
        add_variables = true
        generate_output = 'stress_xx stress_yy stress_zz'
        block = 'BaseMesh Box1'
      []
    []
  []
[]

[AuxVariables]
  [dummy]
    type = MooseVariableFVReal
    block = 'BaseMesh Box1'
  []
[]

# fix the lower model boundary in y and z direction
[BCs]
  [./back_fix_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'back'
    value = 0.0
  []
  [./back_fix_z]
    type = DirichletBC
    variable = disp_z
    boundary = 'back'
    value = 0.0
  []
[]

# fix the left mdoel boundary in x direction
[BCs]
  [./left_fix_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'left'
    value = 0.0
  []
[]

# put some pressure on the right model boundary
[BCs]
  [./right_Dirichlet]
    type = FunctionDirichletBC
    variable = disp_x
    boundary = 'right'
    function = right_pressure_function
  []
[]
[Functions]
  [right_pressure_function]
    type = ParsedFunction
    expression = '-0.001 * t'
  []
[]

# Material: Volume Elements
[Materials]

  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    block = 'BaseMesh Box1'
    youngs_modulus = 1e6
    poissons_ratio = 0.25
  []

  [stress]
    type = ComputeFiniteStrainElasticStress
    block = 'BaseMesh Box1'
  []

[]

# move elements between subdomains back and forth
[UserObjects]
  [GlobalSubdomainModifier]
    type = TimedSubdomainModifier
    times = '0.2'
    blocks_from = 'Box1'
    blocks_to = 'Box1_inactive'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
[]

[Preconditioning]
  [.\SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient

  end_time = 1.0
  dtmin = 0.001
  [TimeSteppers]
    [BlockEventTimeStepper]
      type = TimeSequenceStepper
      time_sequence = '0.05 0.1 0.2 0.4 1.0'
    []
  []

  solve_type = 'PJFNK'
  petsc_options = '-snes_converged_reason'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = ' lu       mumps'

  l_tol = 1E-10
  l_max_its = 20

  nl_abs_tol = 1E-10
  nl_rel_tol = 1e-8
  nl_max_its = 20

[]

[Outputs]
  exodus = true
[]

```

</p>
</details> 

<details><summary> Moose Model with MatDiffusion-Kernels in the inactive subdomain - working </summary>
<p>

```
# This test contains blocks of 3D elements and blocks with 2D elements.

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  use_displaced_mesh = false
[]

Box1_inactive_name = 'Box1_inactive'
inactive_domain_block_names = '${Box1_inactive_name}'

[Problem]
  solve = true
  kernel_coverage_check = SKIP_LIST
  kernel_coverage_block_list = '${inactive_domain_block_names}'
  material_coverage_check = SKIP_LIST
  material_coverage_block_list = '${inactive_domain_block_names}'
[]

[Mesh]
  [BaseMesh]
    type = GeneratedMeshGenerator
    subdomain_name = 'BaseMesh'
    elem_type = 'HEX8'
    dim = 3
    nx = 6
    ny = 6
    nz = 2
    xmin = -3
    xmax = +3
    ymin = -3
    ymax = +3
    zmin = -2
    zmax = +2
  []

  [Box1]
    type = SubdomainBoundingBoxGenerator
    block_name = 'Box1'
    input = "BaseMesh"
    block_id = 1
    location = "INSIDE"
    bottom_left = "-1.0 -1.0 -0.0"
    top_right = "+1.0 +1.0 +2.0"
  []

  add_subdomain_names = '${inactive_domain_block_names}'
[]

[Physics]
  [SolidMechanics]
    [QuasiStatic]
      [all]
        strain = SMALL
        incremental = true
        add_variables = true
        generate_output = 'stress_xx stress_yy stress_zz'
        block = 'BaseMesh Box1'
      []
    []
  []
[]

# ===== Dummy-Kernels for Inactive Domain =====
[Kernels]
  [./donothing1]
    type = MatDiffusion
    block = '${inactive_domain_block_names}'
    variable = disp_z
    diffusivity = 1e-7
  []
 [./donothing2]
    type = MatDiffusion
    block = '${inactive_domain_block_names}'
    variable = disp_x
    diffusivity = 1e-7
  []
 [./donothing3]
    type = MatDiffusion
    block = '${inactive_domain_block_names}'
    variable = disp_y
    diffusivity = 1e-7
  []
[]

[AuxVariables]
  [dummy]
    type = MooseVariableFVReal
  []
[]

# fix the lower model boundary in y and z direction
[BCs]
  [./back_fix_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'back'
    value = 0.0
  []
  [./back_fix_z]
    type = DirichletBC
    variable = disp_z
    boundary = 'back'
    value = 0.0
  []
[]

# fix the left mdoel boundary in x direction
[BCs]
  [./left_fix_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'left'
    value = 0.0
  []
[]

# put some pressure on the right model boundary
[BCs]
  [./right_Dirichlet]
    type = FunctionDirichletBC
    variable = disp_x
    boundary = 'right'
    function = right_pressure_function
  []
[]
[Functions]
  [right_pressure_function]
    type = ParsedFunction
    expression = '-0.001 * t'
  []
[]

# Material: Volume Elements
[Materials]

  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    block = 'BaseMesh Box1'
    youngs_modulus = 1e6
    poissons_ratio = 0.25
  []

  [stress]
    type = ComputeFiniteStrainElasticStress
    block = 'BaseMesh Box1'
  []

[]

# move elements between subdomains back and forth
[UserObjects]
  [GlobalSubdomainModifier]
    type = TimedSubdomainModifier
    times = '0.2'
    blocks_from = 'Box1'
    blocks_to = 'Box1_inactive'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
[]

[Preconditioning]
  [.\SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient

  end_time = 1.0
  dtmin = 0.001
  [TimeSteppers]
    [BlockEventTimeStepper]
      type = TimeSequenceStepper
      time_sequence = '0.05 0.1 0.2 0.4 1.0'
    []
  []

  solve_type = 'PJFNK'
  petsc_options = '-snes_converged_reason'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = ' lu       mumps'

  l_tol = 1E-10
  l_max_its = 20

  nl_abs_tol = 1E-10
  nl_rel_tol = 1e-8
  nl_max_its = 20

[]

[Outputs]
  exodus = true
[]
```

</p>
</details>

**Comment by GiudGiud** [comment, marked as answer]
> How can I check which variables and kernels are active on which subdomains?

One way is to use Debug/show_execution_order and see what kernel is active on what subdomain

> The variables disp_x, disp_y, and disp_z seem to be present in my inactive domain. Is there my problem? If yes, should I block-restrict them as well?

Absolutely.
If they are present on domains that there is no equation for them, then the nullspace is the entire space, and most solvers will not like this.

**Reply by jmeier** [reply]
Thanks for pointing to Debug/show_execution_order. If I interpret the output correctly, there are no kernels on block 2 (as expected, since I do not request kernels for this block in the input file). But I dont see block-coverage of the variables?

```
Time Step 3, time = 0.101, dt = 0.001
[DBG] Computing elemental user objects on TIMESTEP_BEGIN
[DBG] Ordering of User Objects on block 0
[DBG] Executing ElementUserObject on TIMESTEP_BEGIN
[DBG] Order of execution:
[DBG] GlobalSubdomainModifier
                             
[DBG] Only user objects active on local element/sides are executed
[DBG] Ordering of User Objects on block 2
[DBG] Executing ElementUserObject on TIMESTEP_BEGIN
[DBG] Order of execution:
[DBG] GlobalSubdomainModifier
                             
[DBG] Only user objects active on local element/sides are executed
[DBG] Beginning elemental loop to compute Residual on LINEAR
[DBG] Ordering of Residual Objects on block BaseMesh (0)
[DBG] Ordering of kernels:
[DBG] TM_bla0 TM_bla1 TM_bla2
[DBG] Ordering of Residual Objects on block Box1_inactive (2)
 0 Nonlinear |R| = 4.150026e+01
```

Digging a little deeper:

If I read the code correctly that adds the displacement variables, Moose does **not** seem to apply the block constraints to the variables created. Therefore, disp_x, disp_y, and disp_z are also active for all my inactive domains. Is this assessment correct?
https://github.com/idaholab/moose/blob/26a0eda9ec737b3d0d9072230756e05e21e2af5a/modules/solid_mechanics/src/physics/QuasiStaticSolidMechanicsPhysics.C#L385-L402

Just from the syntax of the [Physics/SolidMechanics/QuasiStatic]-block I'd assume Moose is only adding this varibales to the subdomains given in the `blocks` parameter:
```
[Physics]
  [SolidMechanics]
    [QuasiStatic]
      [bla]
        strain = SMALL
        incremental = true
        add_variables = true
        generate_output = 'stress_xx stress_yy stress_zz'
        block = 'BaseMesh Box1'
      []
    []
  []
[]
```

Would it make sense to hand-down the block restriction of [Physics/SolidMechanics/QuasiStatic] to the variables it creates?

If the answer to this question is 'yes' some things are unclear to me. Please assume a hypothetical model:
- With a [Physics/SolidMechanics/QuasiStatic], disp_x, disp_y and disp_z variables are only created for subdomain “A” (e.g. for 3D-volume elements)
- The user also needs the variables disp_x, disp_y and disp_z and others (rotXYZ variables...) for a subdomain “B” with e.g. shell elements. The user must then create these variables explicitly with [Variables]...[].
- Does this mean that Moose creates 2x the variables disp_x, disp_y and disp_z? Or does Mosse automatically merge these variables?

**Reply by GiudGiud** [reply]
Moose does not automatically merge two variables

And defining them twice won't work because you'll get an error. 

Imo you need to add an option to not define the variables inside the Physics, then perform that part manually

**Reply by jmeier** [reply]
Ok, without changing the code of Moose, I have to fall back to “add_variables = false” and add the disp-variables by hand with "correct" block restriction.

Would the following behavior be desirable if “add_variables = true” is set in [Physics/SolidMechanics/QuasiStatic]?
_[Physics/SolidMechanics/QuasiStatic] checks whether the disp variables are already defined in the model. If not, these variables are added with bock-restriction. If the variables already exist and have a block-restriction, the existing block-restriction from [Physics/SolidMechanics/QuasiStatic] is added._

**Reply by GiudGiud** [reply]
This seems fine to me

**Reply by jmeier** [reply]
Now I got a little problem.

If I define the variable `disp_x` above the [Physics/SolidMechanics/QuasiStatic] block, I can dedect this variable in QuasiStaticSolidMechanicsPhysics.C using

```
if (_problem->hasVariable(disp)) ...
```

The corresponding input file:
```
# ===== Variables: Displacements =====
[Variables]
  [disp_x]
    order = SECOND
    family = LAGRANGE
    block = 'BaseMesh'
  []
  ...
[]

[Physics]
  [SolidMechanics]
    [QuasiStatic]
      [all]
        strain = SMALL
        incremental = true
        add_variables = true
        generate_output = 'stress_xx stress_yy stress_zz'
        block = 'Box1'
      []
    []
  []
[]
```

If I have the [Variables] block _below_ the [Physics/SolidMechanics/QuasiStatic] block, `_problem->hasVariable(disp)` returns false. 

Fun fact: Moose does not complain that add_variables is true and disp_x already exists regardless of where which block is located! I think this is because:
https://github.com/idaholab/moose/blob/468ab73aa23733be91b7e142dd2926aeb230ff53/framework/src/problems/FEProblemBase.C#L2562

One option would be to force the user to define Variables **before** [Physics/SolidMechanics/QuasiStatic]. This would add constraints to the design of the input files. I do not like this option much, but it would ease implementing this issue.

In search of other options: Is there a way to detect and modify variables defined after the [Physics/SolidMechanics/QuasiStatic] block (I'd need to merge the block-restriction of [Physics/SolidMechanics/QuasiStatic])

**Reply by GiudGiud** [reply]
Hello

If a task has to happen before the task that adds the variables of the physics, then in the ...App.C file (SolidMechanics most likely) the dependency between the two tasks should be indicated. 

If you add Debug/show_actions=true you can see the order in which the tasks of each action are performed. 

Within the same task, the order in the input file matters

Guillaume

**Reply by jmeier** [reply]
Dear Guillaume,

I need to come byck to one of my initial questions: 
> How can I check which variables and kernels are active on which subdomains?


For Kernels you gave:
> One way is to use Debug/show_execution_order and see what kernel is active on what subdomain

How about variables? How can I find out in which blocks a variable is defined?

**Reply by GiudGiud** [reply]
We don't have that in moose. 

Feel free to add Debug/show_block_restriction to moose if you need it

**Reply by jmeier** [reply]
The name `show_block_restriction` would suggest to show more than just "what variables are defined per block". 

So the output of `show_block_restriction` may look like:

```
Subdomain 'BaseMesh' (0):
    Variables (3): disp_x, disp_y, disp_z
    AuxVariables (1): dummy_aux_var

Subdomain 'TheOtherBlock' (1):
    Variables (3): disp_x, disp_y, disp_z
    AuxVariables (2): dummy_aux_var, another_aux_var
```

Beside Variables and AuxVariables other objects may be shown.

Would that be in the spirit of a debug option `show_block_restriction`?

**Reply by GiudGiud** [reply]
Yes I think this would make be good to do this for all the different systems in use and block-restricted

Alternatively, the code could detect groups of blocks that are shared between groups of objects. This would limit the output. 
However because this is for debug purposes it's not needed now

@YaqiWang dont you have that in griffin?

**Reply by jmeier** [reply]
I started to implement this debug parameter with PR #28291. The PR contains questions about the implementation. I would be happy to receive any hints.

**Reply by jmeier** [reply]
Dear Guillaume,

Unfortunately, even with #28291 ready for review I'm a bit stuck here. I'd like to come back to another of your comments:

> If a task has to happen before the task that adds the variables of the physics, then in the ...App.C file (SolidMechanics most likely) the dependency between the two tasks should be indicated.

In the context of [Physics/SolidMechanics/QuasiStatic].add_variables=true and manually (partially) defined variables: 

1. I think I understood that if I want to enable [Physics/SolidMechanics/QuasiStatic] to modify the block restriction of user-defined variables, all variables must have been created by Moose first. Is this right? So is it not too late to modify the block restriction of the variable then?
2. To define the dependeny needed here I have to do something like here (but 2x with `add_variable`)?
https://github.com/idaholab/moose/blob/b4f5a8ad1c99a084d3df6caa16693e433efc29cf/modules/contact/src/base/ContactApp.C#L47
3. If the answer to point 2 is "yes", how do I differentiate between the `add_variable` for 'real' variables and the `add_variable` task of [Physics/SolidMechanics/QuasiStatic]?

Jörg

**Reply by GiudGiud** [reply]
> I think I understood that if I want to enable [Physics/SolidMechanics/QuasiStatic] to modify the block restriction of user-defined variables, all variables must have been created by Moose first. Is this right? So is it not too late to modify the block restriction of the variable then?

I don't know if that is actually an option. Where in QuasiStaticSolidMechanicsPhysics do you see "adding more blocks" ? I only see variables being added (without the block restriction for displacement actually)

>  syntax.addDependency("add_variable", "add_variable"); 

no you can't do that. You would have to create a new task, like "add_disp_variable" and add it to the Physics
But I believe we are not operating on the right pretenses here. Since variables can only be added once, with the right block restriction from the get-go, then I dont see the need to order the tasks adding variables differently. Maybe I m missing something?

**Reply by jmeier** [reply]
Once again, thank you for your detailed answer.

> Where in QuasiStaticSolidMechanicsPhysics do you see "adding more blocks" ? I only see variables being added (without the block restriction for displacement actually)

It is exactly my problem, that Moose adds the disp-variables (or adopts existing disp-variables) without block restriction.

My questions in [my message](https://github.com/idaholab/moose/discussions/28250#discussioncomment-10236124) were aimed at whether and what possibilities there are for [QuasiStaticSolidMechanicsPhysics ] to add its block restriction to a variable defined by the user in a [Variables] block. But I think you already answered this question:

> Since variables can only be added once, with the right block restriction from the get-go, ...

If you dont see any other options, I have to stick with manually defining my disp-variables in [Variables] and leave `add_variables` always `false`.

**Reply by GiudGiud** [reply]
I think I would be in favor of having the Physics add the displacement variables with its block restriction. I dont remember if I said otherwise elsewhere

**Reply by jmeier** [reply]
I created an issue for that: #28307
And a corresponding draft: #28308
